SSR Externals False Positive on Windows

## Describe the bug

Part of the heuristic for determining if a dependency should be externalized for SSR is to compare the "resolved ESM entry point and its default Node entry point". That comparison happens here: https://github.com/vitejs/vite/blob/5095e041deaced2db8fc3c3af504367bc57bb93f/packages/vite/src/node/ssr/ssrExternal.ts#L67
The path `requireEntry` comes from `require.resolve` which is not normalized. On Windows the path `entry` (which is normalized) has POSIX path separators and `requireEntry` has Windows path separators. This leads to a false positive, thinking the package has a cjs distribution which then fails at runtime.

Simply using Vite's path normalization utility on `requireEntry` also doesn't quite work as, at least in my case, the paths end up with slightly different casing: `"C:\..."` vs `"c:\..."`

## Reproduction

https://github.com/rturnq/solid-hackernews
```
npm i
npm run dev
```


## System Info

- `vite` version: 2.0.5
- Operating System: Windows 10.0.18363
- Node version: 15.4.0
- Package manager (npm/yarn/pnpm) and version: npm 7.0.15

## Logs (Optional if provided reproduction)

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: C:\Some\File\Path\solid-hackernews\node_modules\solid-app-router\dist\index.js
require() of ES modules is not supported.
require() of C:\Some\File\Path\solid-hackernews\node_modules\solid-app-router\dist\index.js from C:\Some\File\Path\solid-hackernews\node_modules\vite\dist\node\chunks\dep-e0f09032.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from C:\Some\File\Path\solid-hackernews\node_modules\solid-app-router\package.json.

    at new NodeError (node:internal/errors:278:15)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1125:13)
    at Module.load (node:internal/modules/cjs/loader:973:32)
    at Function.Module._load (node:internal/modules/cjs/loader:813:14)
    at Module.require (node:internal/modules/cjs/loader:997:19)
    at n (C:\Some\File\Path\solid-hackernews\node_modules\jiti\dist\v8cache.js:2:2349)
    at nodeRequire (C:\Some\File\Path\solid-hackernews\node_modules\vite\dist\node\chunks\dep-e0f09032.js:68055:17)
    at ssrImport (C:\Some\File\Path\solid-hackernews\node_modules\vite\dist\node\chunks\dep-e0f09032.js:68013:20)
    at eval (/src/components/nav.tsx:15:31)
    at instantiateModule (C:\Some\File\Path\solid-hackernews\node_modules\vite\dist\node\chunks\dep-e0f09032.js:68041:166)
```

### Describe the bug
I'm experiencing something somewhat similar to the above on macOS. Looking for some pointers; hopefully this is not completely out of topic.

I'm trying to use the [Contentful Javascript SDK](https://www.npmjs.com/package/contentful) like this:
```
<script>
	import {createClient} from 'contentful';
	console.log('createClient', createClient);
</script>
```

### Logs

**Server**
```
sveltekit % npm run dev --

> ~TODO~@0.0.1 dev /Users/angel.meraz/dev/sveltekit
> svelte-kit dev


  SvelteKit v1.0.0-next.109

  local:   http://localhost:3000
  network: not exposed

  Use --host to expose server to other devices on this network


createClient [Function: createClient]
10:12:55 [vite] new dependencies found: contentful, updating...
 > node_modules/contentful-sdk-core/dist/index.es-modules.js:3:9: error: No matching export in "browser-external:os" for import "platform"
    3 │ import { platform, release } from 'os';
      ╵          ~~~~~~~~

 > node_modules/contentful-sdk-core/dist/index.es-modules.js:3:19: error: No matching export in "browser-external:os" for import "release"
    3 │ import { platform, release } from 'os';
      ╵                    ~~~~~~~

10:12:55 [vite] error while updating dependencies:
Error: Build failed with 2 errors:
node_modules/contentful-sdk-core/dist/index.es-modules.js:3:9: error: No matching export in "browser-external:os" for import "platform"
node_modules/contentful-sdk-core/dist/index.es-modules.js:3:19: error: No matching export in "browser-external:os" for import "release"
    at failureErrorWithLog (/Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:1443:15)
    at /Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:1125:28
    at runOnEndCallbacks (/Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:915:63)
    at buildResponseToResult (/Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:1123:7)
    at /Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:1230:14
    at /Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:606:9
    at handleIncomingPacket (/Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:703:9)
    at Socket.readFromStdout (/Users/angel.meraz/dev/sveltekit/node_modules/esbuild/lib/main.js:573:7)
    at Socket.emit (events.js:315:20)
    at Socket.EventEmitter.emit (domain.js:467:12)
createClient [Function: createClient]
```

**Client**
```
The requested module '/node_modules/contentful/dist/es-modules/contentful.js' does not provide an export named 'createClient'

SyntaxError: The requested module '/node_modules/contentful/dist/es-modules/contentful.js' does not provide an export named 'createClient'
```

As you can see the package is being correctly interpreted during SSR but not on the client.

### Contentful distributed code as an ES module
From my local `/node_modules/contentful/dist/es-modules/contentful.js`.
```
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createClient = createClient;

var _axios = _interopRequireDefault(require("axios"));

var _contentfulSdkCore = require("contentful-sdk-core");

var _createContentfulApi = _interopRequireDefault(require("./create-contentful-api"));

var _createGlobalOptions = _interopRequireDefault(require("./create-global-options"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Create a client instance
 * @func
 * @name createClient
 * @memberof contentful
 * @param {Object} params - Client initialization parameters
 * @prop {string} params.space - Space ID
 * @prop {string} params.accessToken - Contentful CDA Access Token
 * @prop {string} [params.environment="master"] - Contentful Environment ID
 * @prop {boolean=} params.insecure - Requests will be made over http instead of the default https (default: true)
 * @prop {string=} params.host - API host (default: cdn.contentful.com). Also usable with preview.contentful.com.
 * @prop {string=} params.basePath - Path appended to the host to support gateways/proxies with custom urls.
 * @prop {Object=} params.httpAgent - Optional Node.js HTTP agent for proxying (see <a href="https://nodejs.org/api/http.html#http_class_http_agent">Node.js docs</a> and <a href="https://www.npmjs.com/package/https-proxy-agent">https-proxy-agent</a>)
 * @prop {Object=} params.httpsAgent - Optional Node.js HTTP agent for proxying (see <a href="https://nodejs.org/api/http.html#http_class_http_agent">Node.js docs</a> and <a href="https://www.npmjs.com/package/https-proxy-agent">https-proxy-agent</a>)
 * @prop {Object=} params.proxy - Optional Axios proxy (see <a href="https://github.com/mzabriskie/axios#request-config"> axios docs </a>)
 * @prop {Object=} params.headers - Optional additional headers
 * @prop {function=} params.adapter - Optional axios request adapter (see <a href="https://github.com/mzabriskie/axios#request-config"> axios docs </a>)
 * @prop {boolean=?} params.resolveLinks - If we should resolve links between entries (default: true)
 * @prop {boolean=?} params.removeUnresolved - If we should remove links to entries which could not be resolved (default: false)
 * @prop {boolean=?} params.retryOnError - If we should retry on errors and 429 rate limit exceptions (default: true)
 * @prop {function=} params.logHandler - A log handler function to process given log messages & errors. Receives the log level (error, warning & info) and the actual log data (Error object or string). (The default can be found at: https://github.com/contentful/contentful-sdk-core/blob/master/lib/create-http-client.js)
 * @prop {string=?} params.application - Application name and version e.g myApp/version
 * @prop {string=?} params.integration - Integration name and version e.g react/version
 * @prop {number=} params.timeout in milliseconds - connection timeout (default:30000)
 * @prop {number=} params.retryLimit - Optional number of retries before failure. Default is 5
 * @returns {ContentfulClientAPI.ClientAPI}
 * @example
 * const contentful = require('contentful')
 * const client = contentful.createClient({
 *  accessToken: 'myAccessToken',
 *  space: 'mySpaceId'
 * })
 */
function createClient(params) {
  if (!params.accessToken) {
    throw new TypeError('Expected parameter accessToken');
  }

  if (!params.space) {
    throw new TypeError('Expected parameter space');
  }

  const defaultConfig = {
    resolveLinks: true,
    removeUnresolved: false,
    defaultHostname: 'cdn.contentful.com',
    environment: 'master'
  };

  const config = _objectSpread(_objectSpread({}, defaultConfig), params);

  const userAgentHeader = (0, _contentfulSdkCore.getUserAgentHeader)(`contentful.js/${"8.3.6"}`, config.application, config.integration);
  config.headers = _objectSpread(_objectSpread({}, config.headers), {}, {
    'Content-Type': 'application/vnd.contentful.delivery.v1+json',
    'X-Contentful-User-Agent': userAgentHeader
  });
  const http = (0, _contentfulSdkCore.createHttpClient)(_axios.default, config);
  const getGlobalOptions = (0, _createGlobalOptions.default)({
    resolveLinks: config.resolveLinks,
    environment: config.environment,
    removeUnresolved: config.removeUnresolved,
    spaceBaseUrl: http.defaults.baseURL,
    environmentBaseUrl: `${http.defaults.baseURL}environments/${config.environment}`
  }); // Append environment to baseURL

  http.defaults.baseURL = getGlobalOptions().environmentBaseUrl; // Intercepts response and obscure the token

  obscureAuthTokenInResponse(http);
  return (0, _createContentfulApi.default)({
    http,
    getGlobalOptions
  });
}

function obscureAuthTokenInResponse(http) {
  http.interceptors.response.use(response => {
    return response;
  }, error => {
    if (error.response && error.response.config.headers.Authorization) {
      const token = error.response.config.headers.Authorization;
      error.response.config.headers.Authorization = error.response.config.headers.Authorization.replace(token, `Bearer...${token.substr(-5)}`);

      if (error.response.request._headers && error.response.request._headers.authorization) {
        error.response.request._headers.authorization = error.response.request._headers.authorization.replace(token, `Bearer...${token.substr(-5)}`);
      }

      if (error.response.request._header) {
        error.response.request._header = error.response.request._header.replace(token, `Bearer...${token.substr(-5)}`);
      }
    }

    return Promise.reject(error);
  });
}
```

&& 

contentful-sdk-core

https://github.com/contentful/contentful-sdk-core/blob/8adf0edf428b291dbb5e7ba2cb8b5b2cfbbadc8a/src/get-user-agent.ts

--- a/packages/vite/src/node/ssr/ssrExternal.ts
+++ b/packages/vite/src/node/ssr/ssrExternal.ts
@@ -5,6 +5,7 @@ import {
   createDebugger,
   isDefined,
   lookupFile,
+  normalizePath,
   resolveFrom,
   unique
 } from '../utils'
@@ -68,7 +69,9 @@ export function resolveSSRExternal(
         undefined,
         true
       )?.id
-      requireEntry = require.resolve(id, { paths: [root] })
+      // normalizePath required for windows. tryNodeResolve uses normalizePath
+      // which returns with '/', require.resolve returns with '\\'
+      requireEntry = normalizePath(require.resolve(id, { paths: [root] }))
     } catch (e) {
       // resolve failed, assume include
       debug(`Failed to resolve entries for package "${id}"\n`, e)

